src/libraries/System.Private.CoreLib/src/System/Threading/PeriodicTimer.cs

```diff
@@ -0,0 +1,223 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+
+namespace System.Threading
+{
+    /// <summary>Provides a periodic timer that enables waiting asynchronously for timer ticks.</summary>
+    /// <remarks>
+    /// This timer is intended to be used only by a single consumer at a time: only one call to <see cref="WaitForNextTickAsync" />
+    /// may be in flight at any given moment.  <see cref="Stop"/> or the instance's <see cref="IDisposable.Dispose"/> implementation
+    /// may be used concurrently with an active <see cref="WaitForNextTickAsync"/> to interrupt it and cause it to return false.
+    /// </remarks>
+    public sealed class PeriodicTimer : IDisposable
+    {
+        /// <summary>The underlying timer.</summary>
+        private readonly TimerQueueTimer _timer;
+        /// <summary>All state other than the _timer, so that the rooted timer's callback doesn't indirectly root itself by referring to _timer.</summary>
+        private readonly State _state;
+
+        /// <summary>Initializes the timer.</summary>
+        /// <param name="period">The time interval between invocations of callback..</param>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="period"/> must be represent a number of milliseconds larger than 0 and smaller than <see cref="uint.MaxValue"/>.</exception>
+        public PeriodicTimer(TimeSpan period)
+        {
+            long ms = (long)period.TotalMilliseconds;
+            if (ms < 1 || ms > Timer.MaxSupportedTimeout)
+            {
+                GC.SuppressFinalize(this);
+                throw new ArgumentOutOfRangeException(nameof(period));
+            }
+
+            _state = new State();
+            _timer = new TimerQueueTimer(s => ((State)s!).Signal(), _state, (uint)ms, (uint)ms, flowExecutionContext: false);
```

Comment:
Should this start on the first call to `WaitForNextTickAsync` instead?

Reply:
Why?  That seems inconsistent with the timer always running and a WaitForNextTickAsync call being able to complete immediately due to the timer having already fired.  I'm not sure why the first call to it should differ from all the others.

Reply:
I like the idea of there being no resources allocated until there's an operation in flight. It feels more pay for play that way. Also I was looking at ASP.NET Core and we have an explicit call to Start. I can be convinced this isn't as big a deal now because there's no user callback running until the call to `WaitForNextTickAsync` but it feels better.

Reply:
But there are still resources allocated after all operations (if operation is defined as WaitForNextTickAsync()) have quiesced.  The consistent thing then would be to stop the timer when WaitForNextTickAsync completes and start it again the next time it's called.  But that's not the behavior that was agreed upon, so it's strange to me that the first request would differ from the others.